clients/src/main/java/org/apache/kafka/clients/consumer/OffsetResetStrategy.java

```diff
@@ -16,6 +16,40 @@
  */
 package org.apache.kafka.clients.consumer;
 
+import java.util.Locale;
+import java.util.Objects;
+
+/**
+ * @see ConsumerConfig#AUTO_OFFSET_RESET_CONFIG
+ * @see ConsumerConfig#AUTO_OFFSET_RESET_DOC
+ */
 public enum OffsetResetStrategy {
-    LATEST, EARLIEST, NONE
+    LATEST, EARLIEST, NONE;
+
+    // Enums are singletons, this means that this conversion happens once the
+    // first time a variant is actually used, and never again.
+    private final String id = name().toLowerCase(Locale.ROOT);
```

Comment:
Sorry that I can't catch your point here. Could you give more explanation to me? Is it different to ```Enum#toString```?

Reply:
We can move the code to `toString` but than the `toLowerCase` conversion is executed every time `toString` is invoked. This way the conversion happens once: the first time the enum is used. From there on out the string stays and never needs to be recomputed. Meaning:

```java
System.out.println(OffsetResetStrategy.LATEST); // first access, `id` member is populated with `toLowerCase` result
System.out.println(OffsetResetStrategy.LATEST); // just uses `id`
System.out.println(OffsetResetStrategy.LATEST); // just uses `id`
System.out.println(OffsetResetStrategy.LATEST); // just uses `id`
```